vcd/resource_vcd_edgegateway.go

```diff
@@ -206,23 +210,38 @@ func resourceVcdEdgeGatewayRead(d *schema.ResourceData, meta interface{}) error
 
 	vcdClient := meta.(*VCDClient)
 
-	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "name")
+	identifier := d.Id()
+	if identifier == "" {
+		identifier = d.Get("name").(string)
+	}
+	if identifier == "" {
+		return fmt.Errorf("[edgegateway read] no identifier provided")
+	}
+
+	orgName := d.Get("org").(string)
+	vdcName := d.Get("vdc").(string)
+	_, vdc, err := vcdClient.GetOrgAndVdc(orgName, vdcName)
 	if err != nil {
-		d.SetId("")
-		return nil
+		return fmt.Errorf("[edgegateway read] error retrieving org and vdc: %s", err)
+	}
+	edgeGateway, err := vdc.GetEdgeGatewayByNameOrId(identifier, false)
+	if err != nil {
```

Comment:
Do you mean in addition to returning the error? It should be redundant.

Reply:
There is a difference between d.SetId("") and not using it. Without it, one would just bring out an error during read, while if it is set - it would inform terraform that the resource no longer exists.

https://www.terraform.io/docs/extend/writing-custom-providers.html#implementing-read

Reply:
When I use `d.SetId("")` + `return nil`, I get
```
 $ terraform  apply -auto-approve
data.vcd_edgegateway.gw-datacloud: Refreshing state...

Error: "name": required field is not set

  on config.tf line 29, in data "vcd_external_network" "external_network1":
  29: data "vcd_external_network" "external_network1" {
```
Which leaves me baffled about what happened (in this case, I misspelled the entity name).


When I return the error (with or without `d.SetId("")`) I get a more telling reason for the failure:
```
$ terraform  apply -auto-approve
data.vcd_edgegateway.gw-datacloud: Refreshing state...

Error: [edgegateway read] error retrieving edge gateway gw-datacloud-x: [ENF] entity not found

  on config.tf line 23, in data "vcd_edgegateway" "gw-datacloud":
  23: data "vcd_edgegateway" "gw-datacloud" {
```

Reply:
At the moment this follows the pattern of all other resources we have. Breaking this discussion out of this specific PR.